tools/gmhazard_scripts/db_creation/empirical_db/new_calc_emp_ds.py

```diff
@@ -0,0 +1,248 @@
+
+# fmt: off
+PERIOD = [0.01, 0.02, 0.03, 0.04, 0.05, 0.075, 0.1, 0.12, 0.15, 0.17, 0.2, 0.25, 0.3, 0.4, 0.5, 0.6, 0.7, 0.75, 0.8,
+          0.9, 1.0, 1.25, 1.5, 2.0, 2.5, 3.0, 4.0, 5.0, 6.0, 7.5, 10.0]
+# fmt: on
+
+IM_TYPE_LIST = [
+    gc.im.IMType.PGA,
+    gc.im.IMType.pSA,
+    gc.im.IMType.Ds595,
+    gc.im.IMType.Ds575,
+    gc.im.IMType.AI,
+    gc.im.IMType.CAV,
+    gc.im.IMType.PGV,
+]
+
+MAG = [
+    5.25,
+    5.75,
+    6.25,
+    6.75,
+    7.25,
+    8.0,
+]  # Definitions for the maximum magnitude / rjb relation
+DIST = [125, 150, 175, 200, 250, 300]
+
+
+def create_rupture_context_df(
+    distance_df: pd.DataFrame,
+    site_data: pd.Series,
+    nhm_data: pd.DataFrame,
+    tect_type: Enum,
+):
+    """Creating the form of dataframe for the vectorized OQ Wrapper
+    By combining site, distance and rupture information
+    """
+    rupture_df = nhm_data.merge(
+        distance_df, left_on="fault_name", right_on="fault_name"
+    )
+    max_dist = np.minimum(
+        np.interp(rupture_df.mag, MAG, DIST),
+        common.get_max_dist_zfac_scaled(site_data),
+    )
+    rupture_df = rupture_df[rupture_df["rjb"] < max_dist]
+
+    # Adding missing columns
+    # Site Parameters
+    rupture_df["vs30"] = site_data.vs30
+    rupture_df["vs30measured"] = site_data.get("vs30measured", default=False)
+    rupture_df["z1pt0"] = site_data.z1p0
+    rupture_df["z2pt5"] = site_data.z2p5
+
+    # Rupture Parameter
+    # hypo_depth is not dbot but can be used as a proxy for point source
+    # if there is only a single point then the dtop, dbot and hypo_depth
+    # are at the same point
+    rupture_df[["hypo_depth", "ztor"]] = rupture_df[["dbot", "dtop"]]
+
+    # Distance Parameter - OQ uses ry0 term
+    rupture_df[["rx", "ry0"]] = rupture_df[["rx", "ry"]].fillna(0)
+
+    return rupture_df.loc[rupture_df["tect_type"] == tect_type.name]
+
+
+def calculate_emp_ds(
+    background_sources_ffp: str,
+    site_source_db_ffp: str,
+    vs30_ffp: str,
+    output_dir: str,
+    z_ffp: Optional[str],
+    ims: Optional[List[gc.im.IMType]],
+    psa_periods: Optional[List[float]],
+    model_dict_ffp: Optional[str],
+    suffix: Optional[str] = None,
+):
+    nhm_data = gc.utils.ds_nhm_to_rup_df(background_sources_ffp)
+    rupture_df = pd.DataFrame(nhm_data["rupture_name"])
+    model_dict = empirical_factory.read_model_dict(model_dict_ffp)
+
+    with gc.dbs.SiteSourceDB(site_source_db_ffp) as distance_store:
+        # TODO: simplify the get_work() - removing MPI
+        fault_df, _, site_df, _ = common.get_work(
+            distance_store, vs30_ffp, z_ffp, None, None
+        )
+
+        # Check to see if any site's Z1.0 or Z2.5 is NaN
+        if np.any(np.isnan(site_df.z1p0.values)):
+            raise ValueError("Z1.0 cannot be NaN")
+        if np.any(np.isnan(site_df.z2p5.values)):
+            raise ValueError("Z2.5 cannot be NaN")
+
+        # Filter the site/station that the distance_store actually has
+        sites = [
```

Comment:
I suggest getting the available stations from the `distance_store` (i.e. using `stored_stations`) and then using `np.isin` (or the pandas version). For larger number of stations using `np.isin` is significantly faster